StreamDeckPlugin/Services/DynamicActionInfoStore.cs

```diff
@@ -21,39 +20,164 @@ public class DynamicActionInfoStore : IDynamicActionInfoStore {
+            lock (_cacheLock) {
+                dynamicActionInfoWithImageItems = (from dynamicActionInfo in _dynamicActionInfoList
+                                                  where dynamicActionInfo.ImageId == imageId
+                                                  select dynamicActionInfo).ToList();
+            }
 
             foreach (var dynamicActionInfo in dynamicActionInfoWithImageItems) {
                 _eventBus.PublishDynamicActionInfoChanged(dynamicActionInfo);
             }
         }
+
+        private void CardTemplateVisibilityChanged(string name, bool isVisible) {
+            var changedActionInfoList = new List<DynamicActionInfo>();
+            lock (_cacheLock) {
+                foreach (var dynamicActionInfo in _dynamicActionInfoList) {
+                    if (dynamicActionInfo.ImageId == name && dynamicActionInfo.IsToggled != isVisible) {
+                        dynamicActionInfo.IsToggled = isVisible;
+                        changedActionInfoList.Add(dynamicActionInfo);
+                    }
+                }
+            }
+
+            //don't raise events within a lock
+            foreach (var dynamicActionInfo in changedActionInfoList) {
+                _eventBus.PublishDynamicActionInfoChanged(dynamicActionInfo);
+            }
+        }
+
+        private void ButtonInfoChanged(ButtonInfoChanged eventData) {
+            if (eventData.Action == ChangeAction.Update) {
+                UpdateDynamicActionInfo(eventData, eventData);
+            } else {
+                AddDynamicActionInfo(eventData, eventData);
+            }
+        }
+
+        private void ButtonRemoved(ButtonRemoved e) {
+            var changedActionInfoList = new List<DynamicActionInfo>();
+            lock (_cacheLock) {
+                var dynamicActionToRemove = _dynamicActionInfoList.FirstOrDefaultWithContext(e);
+                //changedActionInfoList.Add(dynamicActionToRemove);
+                _dynamicActionInfoList.Remove(dynamicActionToRemove);
+
+                var lastIndex = dynamicActionToRemove.Index;
+                foreach (var dynamicActioninfo in _dynamicActionInfoList) {
+                    if (dynamicActioninfo.IsAfter(e)) {
+                        if (dynamicActioninfo.Index > lastIndex) {
+                            lastIndex = dynamicActioninfo.Index;
+                        }
+                        dynamicActioninfo.Index--;
+                        changedActionInfoList.Add(dynamicActioninfo);
+                    }
+                }
+
+                //we have to add a blank action info at the end so we blank out the previously last item (the list is now shorter)
+                var blankDynamicAction = new DynamicActionInfo(e) {
+                    Index = lastIndex
+                };
+                changedActionInfoList.Add(blankDynamicAction);
+            }
+
+            //don't raise events within a lock
```

Comment:
Nit: This comment is repeated unnecessarily

Reply:
I was trying to remind myself to be careful!
(done)